### PR TITLE
Update sound searcher and spawner for load handling.

### PR DESCRIPTION
**Pull request by howard-stearns**
Less load, more predictable behavior.

### CODE REVIEW

**Comment by howard-stearns** [comment]
https://app.asana.com/0/32622044445063/91007520183310

**Comment by howard-stearns** [comment]
Testing notes: Using the updated examples/audioExamples/acAudioSearching/acAudioSearchCompatibleEntitySpawner.js, one should be able to create a large number of entities (12k) of which many have audio data (2k). The ac script should be able to handle this without (crashing or) falling too far behind. The server log will periodically report how many entities it sees, how many with sounds, how many are playing (should never be more than 40 with todays audio library), and how long between updates (which should not be too far behind intended time of 100 ms).

And of course, you should hear the audio - 40 tents flapping randomly in the breeze.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4bb76d6a203047a36ce15f8d1592e4361208cd84 available at: https://gist.github.com/dd272eac75f79cfbd10f<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7177.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7177.dmg

**Comment by ericrius1** [comment]
Test scripts:
https://rawgit.com/howard-stearns/hifi/ac-sound-searcher-update/examples/audioExamples/acAudioSearching/acAudioSearchCompatibleEntitySpawner.js

https://rawgit.com/howard-stearns/hifi/ac-sound-searcher-update/examples/audioExamples/acAudioSearching/ACAudioSearchAndInject.js

**Comment by thoys** [comment]
## Test report

- I hear many sails flapping, no problems, couldn't hear the tents.
- I have the script running for over 5 minutes.
- The assignment-client that searches for the audio uses 37% of my CPU.
- Memory usage for this assignment-client doesn't grow (No memory leaks)


One little thing I bumped into, is that you cannot run acAudioSearchCompatibleEntitySpawner.js as an assignment script. Because it makes calls to Camera.

One other minor detail is that while Im running this my log has grown to 35 mb:
``` 
[02/24 02:20:27] [DEBUG] [7960] [agent] AudioInjectorManager::threadInjector could not thread AudioInjector - at max of 40 current audio injectors.
[02/24 02:20:27] [DEBUG] [7960] [agent] AudioInjectorManager::threadInjector could not thread AudioInjector - at max of 40 current audio injectors.
[02/24 02:20:27] [DEBUG] [7960] [agent] AudioInjectorManager::threadInjector could not thread AudioInjector - at max of 40 current audio injectors.
[02/24 02:20:27] [DEBUG] [7960] [agent] AudioInjectorManager::threadInjector could not thread AudioInjector - at max of 40 current audio injectors.
[02/24 02:20:27] [DEBUG] [7960] [agent] AudioInjectorManager::threadInjector could not thread AudioInjector - at max of 40 current audio injectors.
[02/24 02:20:27] [DEBUG] [7960] [agent] AudioInjectorManager::threadInjector could not thread AudioInjector - at max of 40 current audio injectors.
[02/24 02:20:27] [DEBUG] [7960] [agent] AudioInjectorManager::threadInjector could not thread AudioInjector - at max of 40 current audio injectors.
[02/24 02:20:27] [DEBUG] [7960] [agent] AudioInjectorManager::threadInjector could not thread AudioInjector - at max of 40 current audio injectors.
[02/24 02:20:27] [DEBUG] [7960] [agent] AudioInjectorManager::threadInjector could not thread AudioInjector - at max of 40 current audio injectors.
[02/24 02:20:27] [DEBUG] [7960] [agent] AudioInjectorManager::threadInjector could not thread AudioInjector - at max of 40 current audio injectors.
[02/24 02:20:27] [DEBUG] [7960] [agent] AudioInjectorManager::threadInjector could not thread AudioInjector - at max of 40 current audio injectors.
[02/24 02:20:27] [DEBUG] [7960] [agent] AudioInjectorManager::threadInjector could not thread AudioInjector - at max of 40 current audio injectors.
[02/24 02:20:27] [DEBUG] [7960] [agent] AudioInjectorManager::threadInjector could not thread AudioInjector - at max of 40 current audio injectors.
[02/24 02:20:27] [DEBUG] [7960] [agent] AudioInjectorManager::threadInjector could not thread AudioInjector - at max of 40 current audio injectors.
[02/24 02:20:27] [DEBUG] [7960] [agent] AudioInjectorManager::threadInjector could not thread AudioInjector - at max of 40 current audio injectors.
[02/24 02:20:27] [DEBUG] [7960] [agent] AudioInjectorManager::threadInjector could not thread AudioInjector - at max of 40 current audio injectors.
[02/24 02:20:27] [DEBUG] [7960] [agent] AudioInjectorManager::threadInjector could not thread AudioInjector - at max of 40 current audio injectors.
[02/24 02:20:27] [DEBUG] [7960] [agent] AudioInjectorManager::threadInjector could not thread AudioInjector - at max of 40 current audio injectors.
``` 

I managed to find the number of entities in the logfile aswell:

```
[02/24 02:24:22] [DEBUG] [7960] [agent] script:print()<< {"entities":6784,"sounds":1117,"playing":40,"millisecondsPerUpdate":139.66}
```


While writing this report I still hear the cloth sails.

But the most import part is that this works! Nice job guys!

67cc66dc947d6a848734805f1be85a67

**Comment by howard-stearns** [comment]
Thanks @thoys and @ericrius1. I guess the numbers crunch out to have a an awful lot of sounds ready to play each second, even though only a few will be able to. On average the number we try and fail should be (nSounds - nPlaying) / (soundDuration + gap in seconds) = nearly 400 each second. I think that would explain the excessive failure logging and falling behind (because we can generally only handle a few hundred getEntityProperties per second). Let's do an experiment and push the playbackGap in the spawner out to 7 seconds. If I'm right, the log should be something approaching 40-50% smaller, and we shouldn't be much behind. (My wife won't let me balance the checkbook, so I probably don't have the math right...)

**Comment by hifi-gustavo** [comment]
Build console output for commit: d1088031bfbd024c42990dabad7c22dc197b1534 available at: https://gist.github.com/dd272eac75f79cfbd10f<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7177.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7177.dmg

**Comment by howard-stearns** [comment]
DNM because I'm going to update the source with the math for what to expect in the logs, and verify that we're seeing the right thing.

**Comment by howard-stearns** [comment]
ok, that should do it

**Comment by hifi-gustavo** [comment]
Build console output for commit: 274f72af52994b6d008aab9460f2399ac26769d6 available at: https://gist.github.com/dd272eac75f79cfbd10f<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7177.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7177.dmg

File: examples/audioExamples/acAudioSearching/ACAudioSearchAndInject.js [hunk 1]
```diff
@@ -1,218 +1,258 @@
...
+        ensureSoundData(); // We'll try to play/setOptions and will need position, so we might as well get soundData, too.
+        if (soundData.url !== that.url) {                                    // WAITING => NO DATA (update next time around)
+            return that.stop();
+        }
+        var options = {
+            position: properties.position,
+            loop: soundData.loop || DEFAULT_SOUND_DATA.loop,
+            volume: soundData.volume || DEFAULT_SOUND_DATA.volume
+        };
+        function repeat() { return !options.loop && (soundData.playbackGap >= 0); }
```
Comment (ZappoMan) [inline examples/audioExamples/acAudioSearching/ACAudioSearchAndInject.js]: coding standard says no to this...

**Comment by howard-stearns** [comment]
@thoys - Thanks for looking at this. In response, I did take a look at the machinery. I determined that the logging was correct -- we do want to be noisy about someone using this in an environment in which there are lots of sounds being asked for that aren't playing -- but there was indeed a bug that I wouldn't have noticed if I hadn't looked. I've corrected that now.

Also, the spawner script isn't intended as an ac script. (Bad name.) It's a test/utility to be run only as an interface script (as you saw).

Can we mark this QA passed now?

**Comment by hifi-gustavo** [comment]
Build console output for commit: 077d3310049aadd7d1e2986f57672d47e71d9f05 available at: https://gist.github.com/dd272eac75f79cfbd10f<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7177.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7177.dmg

**Comment by cozza13** [comment]
checked and marked it QA approved

**Pull request merged**